contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3661,101 +3662,151 @@ get_insert_bulk_kilobytes_per_batch()
 	 */
 	login_is_db_owner = 0 == strncmp(login, get_owner_of_db(dbname), NAMEDATALEN);
 	datdba = get_role_oid("sysadmin", false);
 	schema_name = get_physical_schema_name(dbname, stmt->schema_name);
-	schemaOid = LookupExplicitNamespace(schema_name, true);
+	if(schema_name)
+	{
+		schemaOid = LookupExplicitNamespace(schema_name, true);
+	}
+	else
+	{
+		ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_SCHEMA),
+					 errmsg("An object or column name is missing or empty. For SELECT INTO statements, verify each column has a name. For other statements, look for empty alias names. Aliases defined as \"\" or [] are not allowed. Change the alias to a valid name.")));
+	}
 
 	if (!OidIsValid(schemaOid))
 			ereport(ERROR,
 					(errcode(ERRCODE_UNDEFINED_SCHEMA),
 					 errmsg("schema \"%s\" does not exist",
 							schema_name)));
-	
-	if (!is_member_of_role(GetSessionUserId(), datdba) && !login_is_db_owner && !pg_namespace_ownercheck(schemaOid, GetUserId()))
-		ereport(ERROR,
-				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-				 errmsg("Cannot find the schema \"%s\", because it does not exist or you do not have permission.", stmt->schema_name)));
 
-	foreach(lc1, stmt->privileges)
+	foreach(lc, stmt->grantees)
 	{
-		char	*priv_name = (char *) lfirst(lc1);
-		foreach(lc, stmt->grantees)
-		{
-			char	   *grantee_name = (char *) lfirst(lc);
-			Oid	role_oid;
-			bool		grantee_is_db_owner;
+		char		*grantee_name = (char *) lfirst(lc);
+		char	*user = GetUserNameFromId(GetUserId(), false);
+		Oid	role_oid;
+		int16 old_priv = 0;
+		if(strcmp(grantee_name, "public") != 0)
 			rolname	= get_physical_user_name(dbname, grantee_name);
-			role_oid = get_role_oid(rolname, true);
-			grantee_is_db_owner = 0 == strncmp(grantee_name, get_owner_of_db(dbname), NAMEDATALEN);
-
+		else
+			rolname = "public";
+		role_oid = get_role_oid(rolname, true);
 
-			if (role_oid == InvalidOid)
-				ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-					errmsg("Cannot find the principal '%s', because it does not exist or you do not have permission.", grantee_name)));
+		if (strcmp(grantee_name, "public") != 0 && role_oid == InvalidOid)
+			ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				errmsg("Cannot find the principal '%s', because it does not exist or you do not have permission.", grantee_name)));
 
-			if (pg_namespace_ownercheck(schemaOid, role_oid) || is_member_of_role(role_oid, datdba) || grantee_is_db_owner)
-				ereport(ERROR,
-					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-					 errmsg("Cannot grant, deny, or revoke permissions to sa, dbo, entity owner, information_schema, sys, or yourself.")));
+		if ((strcmp(rolname, user) == 0) || pg_namespace_ownercheck(schemaOid, role_oid) || is_member_of_role(role_oid, datdba))
+			ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					errmsg("Cannot grant, deny, or revoke permissions to sa, dbo, entity owner, information_schema, sys, or yourself.")));
```

Comment:
What does 'yourself' mean here?

Reply:
Is this the same as SQL Server error msg?

Reply:
Yourself means when the grantee is the current user itself.
Yes, the error message is same as that of SQL Server.